### ISSUE TITLE
Trackpad pointer and enter button are not working when xremap is running

### ISSUE THREAD

**Issue by danyspin97** [k0kubun/xremap#105]
I have this weird bug when I run xremap: both the trackpad pointer and the enter button stop working. I have tried with both sudo and without it, and the result is the same. I can confirm that xremap reports that only the device it is capturing is the keyboard. As soon as I stop xremap the enter button start working again. While for the mouse pointer, it usually start working again when I open new windows. I have also tried with `--device` but with no luck. I am using sway on openSUSE Tumbleweed currently.

I have tried plugging another external keyboard, and the enter buttons stop working for each keyboard, regardless of `--device` selected.

**Comment by k0kubun** [comment]
Just to confirm, both the trackpad pointer and the enter button are on your keyboard that xremap deals with, is that right?

Could you try starting xremap with `RUST_LOG=debug` like `RUST_LOG=debug xremap` and share the entire command output of xremap while using the trackpad pointer and the enter button? I suspect these are events that xremap might not be handling well, and I need to see what events they are.

**Comment by danyspin97** [comment]
> Just to confirm, both the trackpad pointer and the enter button are on your keyboard that xremap deals with, is that right?

They are not. And when I have plugged in a second keyboard, I have chosen just one with `--device`. The enter button stopped working on all keyboards.

This is the output:
[github-output.txt](https://github.com/k0kubun/xremap/files/8781482/github-output.txt)

The ENTER button is registered, but it's not forwarded to the system. I could see in the terminal the other characters written.

About the trackpad, could it be some problem when it gets opened via `open` with write mode?

**Comment by k0kubun** [comment]
Thank you for providing the output.

> About the trackpad, could it be some problem when it gets opened via open with write mode?

xremap doesn't interfere with devices not selected by `--device`. If some behavior of devices not selected by xremap changes, it's likely due to other processes or drivers. 

What xremap does is:

1. "grab"s event devices selected by `--device`. Events generated by the devices will only go to the xremap process.
2. xremap reads events from the devices and writes replicated/remapped events to a virtual uinput device.

On 2, it's possible that xremap filters out events from devices, but it could happen only for devices grabbed by xremap. 

---

One possibility is that creating a uinput device itself triggers something that impacts all devices. To test it, could you run `xremap --device a_string_that_matches_no_device --watch` and see if the issue reproduces? xremap stops by default when it selects no device, but `--watch` allows it to keep waiting for such a device to appear.

**Comment by danyspin97** [comment]
Thanks for the detailed explanation.

> To test it, could you run xremap --device a_string_that_matches_no_device --watch and see if the issue reproduces?

It does not reproduce, the pointer works fine.

I have used `cat` to check which input device handle the pointer, and it's not the device grabbed by xremap. It's `/dev/input/event2` (xremap only grabs `/dev/input/event0` on my system).

By reading the source code, it seems to me that the logic is slightly different when `--device` is used, but maybe it's not relevant to this issue.

To recap, the issue is that the pointer (`/dev/input/event2`) stops working when xremap starts and successfully grab the keyboard (`/dev/input/event0`). This doesn't happen when xremap does not grab any device. I can also confirm that `/dev/input/event2` is still sending events when xremap has grabbed a device. The issue is that sway isn't receiving them at all.

**Comment by k0kubun** [comment]
A few more questions:

* Have you been using `--watch` since before I asked you? If so, can you test the same thing without `--watch`?
* What happens if you select both "AT Translated Set 2 keyboard" and "ETPS/2 Elantech Touchpad" with `--device` at the same time?
* Do you have an external mouse? Could you reproduce the same thing with the device?
* Can you reproduce the same thing on another distribution, e.g. Ubuntu 22.04?

I only have external keyboards and an external mouse. Unless you tell me how to reproduce your problem on my machine with such devices, I cannot start working on this. This issue is still in a stage waiting for you to explain how to reproduce the problem. Note that I have a Sway environment on the same machine but it doesn't reproduce with an external mouse. If you can prepare a VirtualBox image that reproduces the problem with an external mouse, for example, I can start looking into it.

**Comment by k0kubun** [comment]
I might have successfully reproduced a similar issue on the VirtualBox image provided at https://github.com/k0kubun/xremap/issues/100#issuecomment-1140051545. It seems like I cannot use an enter key when xremap is running. The enter key also randomly stops working even when xremap is not running, so the environment itself is a bit suspicious though. Anyway, I can at least look into that issue in the environment, hoping that fixing it will also fix your problem.

**Comment by Anarchanas** [comment]
I can reproduce the behavior. When starting xremap the enter key is hitted a lot of times and fills the whole terminal. Also the trackpoint is not working anymore even after killing xremap. However when I run it with --device touchpad it starts working again. 
When I select both devices the problem is the same as when just choosing the translated keyboard. 
 I am on debian 11 gnome wayland. 
 
![image](https://user-images.githubusercontent.com/32162309/171675402-890dfacf-8c77-4009-b005-529d412c4017.png)

**Comment by danyspin97** [comment]
> Have you been using --watch since before I asked you? If so, can you test the same thing without --watch?

Yes, if I use watch it continues running even without selecting a device. The issue discussed only happen when xremap has a device selected (either automatically or via --device).

**Comment by k0kubun** [comment]
Thank you for confirming it. I'll be busy in the coming few weeks, but I plan to work on fixing it using the VM that reproduced the issue after that. I appreciate your patience.

**Comment by danyspin97** [comment]
Wev cannot report any input from the mouse either. I ran `swaymsg -t get_input` and I got the following output:

```
Input device: xremap pid=7102
  Type: Mouse
  Identifier: 4660:22136:xremap_pid=7102
  Product ID: 22136
  Vendor ID: 4660
  Libinput Send Events: enabled

Input device: xremap pid=7102
  Type: Keyboard
  Identifier: 4660:22136:xremap_pid=7102
  Product ID: 22136
  Vendor ID: 4660
  Active Keyboard Layout: English (US)
  Libinput Send Events: enabled
```

Is the mouse supposed to be created?

**Comment by k0kubun** [comment]
I haven't reproduced any mouse-related issues, so I can't help you with that until any of you provide a VM that reproduces it. My plan is to fix the enter key issue first and see if any other problem persists.

**Comment by danyspin97** [comment]
@Anarchanas can you please share the permissions and owner/group of `/dev/input/` folder and `/dev/uinput`?

**Comment by akhiljalagam** [comment]
I got the same issue. but when i used watch command, disconnected and connected keyboard usb cable. and then enter key is working again.


logs:

> ./xremap --watch xremap.yml

```
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
/dev/input/event0 : AT Translated Set 2 keyboard
/dev/input/event1 : Keychron Keychron K6
------------------------------------------------------------------------------
Found a removed device. Reselecting devices.
Failed to ungrab device 'Keychron Keychron K6' at '/dev/input/event1' due to: No such device (os error 19)
Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : AT Translated Set 2 keyboard
/dev/input/event10: DELL07EA:00 06CB:7E92 Mouse
/dev/input/event11: DELL07EA:00 06CB:7E92 Touchpad
/dev/input/event12: Video Bus
/dev/input/event13: Video Bus
/dev/input/event14: DP-1
/dev/input/event15: Integrated_Webcam_HD: Integrate
/dev/input/event16: Integrated_Webcam_HD: Integrate
/dev/input/event17: ELAN2097:00 04F3:2541
/dev/input/event18: ELAN2097:00 04F3:2541 UNKNOWN
/dev/input/event19: ELAN2097:00 04F3:2541 UNKNOWN
/dev/input/event2 : Keychron Keychron K6
/dev/input/event20: ELAN2097:00 04F3:2541 Stylus
/dev/input/event21: HDA Intel PCH Headphone Mic
/dev/input/event22: HDA Intel PCH HDMI/DP,pcm=3
/dev/input/event23: HDA Intel PCH HDMI/DP,pcm=7
/dev/input/event24: HDA Intel PCH HDMI/DP,pcm=8
/dev/input/event25: HDA Intel PCH HDMI/DP,pcm=9
/dev/input/event26: HDA Intel PCH HDMI/DP,pcm=10
/dev/input/event27: PS/2 Synaptics TouchPad
/dev/input/event28: xremap pid=22130
/dev/input/event3 : PixArt Dell MS116 USB Optical Mouse
/dev/input/event4 : Lid Switch
/dev/input/event5 : Power Button
/dev/input/event6 : Sleep Button
/dev/input/event7 : Power Button
/dev/input/event8 : Intel HID events
/dev/input/event9 : Dell WMI hotkeys
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
/dev/input/event0 : AT Translated Set 2 keyboard
------------------------------------------------------------------------------
/dev/input/event1 : Keychron Keychron K6
Failed to grab device 'Keychron Keychron K6' at '/dev/input/event1' due to: Device or resource busy (os error 16)
Failed to grab device 'Keychron Keychron K6' at '/dev/input/event1' due to: Device or resource busy (os error 16)
```

**Comment by k0kubun** [comment]
Looking at the ungrab/grab errors, I'm not sure if it's the same issue as the one I reproduced, but the current state is still the same as https://github.com/k0kubun/xremap/issues/105#issuecomment-1146734058. I cannot work on this until I get to.

**Comment by Anarchanas** [comment]
> @Anarchanas can you please share the permissions and owner/group of `/dev/input/` folder and `/dev/uinput`?

Sorry for some reason I did not get no notification so here my late answer

The permissions for /dev/uinput and /dev/input are both root:root

**Comment by rsuu** [comment]
I can not use Enter also.(On Sway)
But this command work for me.
I do not why.

```bash
sleep 1; sudo xremap test.yml
  # work 

sudo xremap test.yml
  # not work
```

config:
```yaml
default_mode: en_US

modmap:
  - name: swap esc and cap
    remap:
      CapsLock: Esc
      Esc: CapsLock

keymap:
  - mode: en_US
    remap:
      Alt-Tab: { set_mode: t }
  - mode: t
    remap:
      Alt-Tab: { set_mode: en_US }
```

**Comment by k0kubun** [comment]
Update: Same as https://github.com/k0kubun/xremap/issues/102, but as I'm currently focusing on a big change for another OSS, I'm planning to work on this issue on Aug 13th. If I receive a pull request before that, I can take a look at that too.

**Comment by k0kubun** [comment]
I took a look at the behavior of evdev on Wayland today. I observed two issues in it, and one of them seems to be causing this issue (and the other is causing #100 and #139).

It seems like a bug in Wayland or at least something that Wayland handles differently from X11. I'll need to work on making a minimum repro and reporting it to the maintainers of Wayland and/or uinput to ask how this could be fixed. As I need to focus on conference preparations for a few weeks, please allow me to work on this after Sep 10th or so.

**Comment by jpzk** [comment]
Also run into the same bug; Would really appreciate a fix- thank you!

**Comment by kevinmost** [comment]
I'm running into this too, and the `sleep 1` workaround suggested by @rsuu is (somehow) working.

If you use a systemctl service to load `xremap` like I do, you can add an `ExecStartPre` to sleep:

```shell
$cat ~/.config/systemd/user/xremap.service

[Unit]
Description=xremap

[Service]
# sleep workaround from https://github.com/k0kubun/xremap/issues/105#issuecomment-1190994137
ExecStartPre=/bin/sleep 1
ExecStart=%h/.cargo/bin/xremap --watch=device %h/.config/xremap/config.yml
Restart=always

[Install]
WantedBy=default.target
```

**Comment by neophrema** [comment]
this sound awfully like the bug I was experiencing, this seems to be related to:
[(https://bugs.freedesktop.org/show_bug.cgi?id=101796)]
As @k0kubun mentioned in https://github.com/k0kubun/xremap/issues/192

I wrote "exec --no-startup-id xremap /home/user/.config/xremap/config.yml" into my swayconfig.
I suspect that as soon as you enter xremap, a constant buttonpress gets registered which adds up to the ghosting.

Try to run the programm as autoload (as @kevinmost concluded also).

**Comment by danyspin97** [comment]
This is a late update because I haven't used my laptop in a lot of time.

I have just updated xremap to the latest version (0.8.12) and this issue has since disappeared. Something in the update fixed it, because it was still reproducible with the old version (0.4.2).

Thank you for the awesome work!!

**Comment by liz-desartiges** [comment]
I'm having the same issue under gnome I built today with cargo and gnome support, the sleep workarround works for me but it's not so convinient

```
$ xremap --version
xremap 0.10.0
```
